File: modern-cms-platform-enhanced/backend/src/controllers/user.controller.js
```javascript
const userService = require('../services/user.service');
const logger = require('../utils/logger');
const { deleteCache } = require('../utils/cache');

// Cache key for all users
const ALL_USERS_CACHE_KEY = '/api/users';

/**
 * Get all users. (Admin only)
 */
exports.getAllUsers = async (req, res, next) => {
    try {
        const users = await userService.findAllUsers();
        res.status(200).json(users);
    } catch (error) {
        logger.error('Error fetching all users:', error.message);
        next(error);
    }
};

/**
 * Get a single user by ID. (Admin only)
 */
exports.getUserById = async (req, res, next) => {
    try {
        const { id } = req.params;
        const user = await userService.findUserById(id);
        if (!user) {
            return res.status(404).json({ message: 'User not found.' });
        }
        res.status(200).json(user);
    } catch (error) {
        logger.error(`Error fetching user by ID ${req.params.id}:`, error.message);
        next(error);
    }
};

/**
 * Update a user. (Admin only)
 */
exports.updateUser = async (req, res, next) => {
    try {
        const { id } = req.params;
        const updatedUser = await userService.updateUser(id, req.body);
        if (!updatedUser) {
            return res.status(404).json({ message: 'User not found.' });
        }
        deleteCache(ALL_USERS_CACHE_KEY); // Invalidate cache
        deleteCache(`/api/users/${id}`); // Invalidate specific user cache
        res.status(200).json({ message: 'User updated successfully.', user: updatedUser });
    } catch (error) {
        logger.error(`Error updating user ${req.params.id}:`, error.message);
        next(error);
    }
};

/**
 * Delete a user. (Admin only)
 */
exports.deleteUser = async (req, res, next) => {
    try {
        const { id } = req.params;
        const success = await userService.deleteUser(id);
        if (!success) {
            return res.status(404).json({ message: 'User not found.' });
        }
        deleteCache(ALL_USERS_CACHE_KEY); // Invalidate cache
        deleteCache(`/api/users/${id}`); // Invalidate specific user cache
        res.status(200).json({ message: 'User deleted successfully.' });
    } catch (error) {
        logger.error(`Error deleting user ${req.params.id}:`, error.message);
        next(error);
    }
};
```

File: modern-auth-system-platform-optimized/backend/src/main/java/com/example/authsystem/controller/UserController.java
```java
package com.example.authsystem.controller;

import com.example.authsystem.dto.UserDto;
import com.example.authsystem.entity.User;
import com.example.authsystem.exception.ResourceNotFoundException;
import com.example.authsystem.service.UserService;
import io.swagger.v3.oas.annotations.Operation;
import io.swagger.v3.oas.annotations.media.Content;
import io.swagger.v3.oas.annotations.media.Schema;
import io.swagger.v3.oas.annotations.responses.ApiResponse;
import io.swagger.v3.oas.annotations.security.SecurityRequirement;
import io.swagger.v3.oas.annotations.tags.Tag;
import jakarta.validation.Valid;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.cache.annotation.CacheEvict;
import org.springframework.cache.annotation.CachePut;
import org.springframework.cache.annotation.Cacheable;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.security.access.prepost.PreAuthorize;
import org.springframework.web.bind.annotation.*;

import java.util.List;
import java.util.UUID;
import java.util.stream.Collectors;

@RestController
@RequestMapping("/api/users")
@RequiredArgsConstructor
@Slf4j
@SecurityRequirement(name = "bearerAuth") // All endpoints in this controller require authentication
@Tag(name = "User Management", description = "Endpoints for managing user profiles (Admin and Self-Service)")
public class UserController {

    private final UserService userService;

    @Operation(summary = "Get all users (Admin only)", responses = {
            @ApiResponse(responseCode = "200", description = "List of users retrieved",
                    content = @Content(mediaType = "application/json", schema = @Schema(implementation = UserDto.class))),
            @ApiResponse(responseCode = "403", description = "Access denied"),
            @ApiResponse(responseCode = "401", description = "Unauthorized")
    })
    @GetMapping
    @PreAuthorize("hasRole('ADMIN')")
    public ResponseEntity<List<UserDto>> getAllUsers() {
        log.info("Admin fetching all users.");
        List<UserDto> users = userService.findAllUsers().stream()
                .map(UserDto::fromEntity)
                .collect(Collectors.toList());
        return ResponseEntity.ok(users);
    }

    @Operation(summary = "Get user by ID (Admin or Self-Service)", responses = {
            @ApiResponse(responseCode = "200", description = "User retrieved",
                    content = @Content(mediaType = "application/json", schema = @Schema(implementation = UserDto.class))),
            @ApiResponse(responseCode = "404", description = "User not found"),
            @ApiResponse(responseCode = "403", description = "Access denied"),
            @ApiResponse(responseCode = "401", description = "Unauthorized")
    })
    @GetMapping("/{id}")
    @PreAuthorize("hasRole('ADMIN') or @userService.isCurrentUser(#id)") // Custom expression for self-service
    @Cacheable(value = "users", key = "#id") // Cache user details
    public ResponseEntity<UserDto> getUserById(@PathVariable UUID id) {
        log.info("Fetching user with ID: {}", id);
        User user = userService.findUserById(id)
                .orElseThrow(() -> new ResourceNotFoundException("User not found with id: " + id));
        return ResponseEntity.ok(UserDto.fromEntity(user));
    }

    @Operation(summary = "Update user by ID (Admin or Self-Service)", responses = {
            @ApiResponse(responseCode = "200", description = "User updated successfully",
                    content = @Content(mediaType = "application/json", schema = @Schema(implementation = UserDto.class))),
            @ApiResponse(responseCode = "404", description = "User not found"),
            @ApiResponse(responseCode = "403", description = "Access denied"),
            @ApiResponse(responseCode = "401", description = "Unauthorized")
    })
    @PutMapping("/{id}")
    @PreAuthorize("hasRole('ADMIN') or @userService.isCurrentUser(#id)")
    @CachePut(value = "users", key = "#id") // Update cache entry
    public ResponseEntity<UserDto> updateUser(@PathVariable UUID id, @Valid @RequestBody UserDto userDto) {
        log.info("Updating user with ID: {}", id);
        User updatedUser = userService.updateUser(id, userDto);
        return ResponseEntity.ok(UserDto.fromEntity(updatedUser));
    }

    @Operation(summary = "Delete user by ID (Admin only)", responses = {
            @ApiResponse(responseCode = "204", description = "User deleted successfully"),
            @ApiResponse(responseCode = "404", description = "User not found"),
            @ApiResponse(responseCode = "403", description = "Access denied"),
            @ApiResponse(responseCode = "401", description = "Unauthorized")
    })
    @DeleteMapping("/{id}")
    @PreAuthorize("hasRole('ADMIN')")
    @ResponseStatus(HttpStatus.NO_CONTENT)
    @CacheEvict(value = "users", key = "#id") // Evict user from cache
    public ResponseEntity<Void> deleteUser(@PathVariable UUID id) {
        log.info("Admin deleting user with ID: {}", id);
        userService.deleteUser(id);
        return ResponseEntity.noContent().build();
    }
}
```